### ISSUE TITLE
Portable pointer initialisation

### ISSUE THREAD

**Issue by elfring** [libimobiledevice/libusbmuxd#14]
I suggest to [reconsider an implementation detail for pointers](http://bytes.com/topic/c/answers/818953-pointer-memset-zero-necessarily-eqal-null).

```
 void collection_init(struct collection *col)
 {
         col->list = malloc(sizeof(void *));
-        memset(col->list, 0, sizeof(void *));
+        col->list = 0;
         col->capacity = 1;
 }
```

Would you like to integrate [this update suggestion](http://c-faq.com/null/null2.html) into your source code repository?

**Comment by nikias** [comment]
I don't think this is correct. It will overwrite the allocation address with a null pointer instead of writing the null pointer to the allocated memory location.
It should be
```
col->list[0] = NULL;
```

**Comment by elfring** [comment]
Are you going to fix the pointer (array) initialisation?

**Comment by nikias** [comment]
I was about to, but then also I was thinking of having an initial capacity of 8 pointers. In that case a memset with `sizeof(void*) * 8` looks better than a `for(i=0;i<8;i++) { col->list[i] = NULL; }`
Also then what about this, it would also have to be changed:
https://github.com/libimobiledevice/libusbmuxd/blob/master/common/collection.c#L55
I don't see any issues with the original approach.

**Comment by elfring** [comment]
>I don't see any issues with the original approach.

* I suggest to use portable pointer assignments instead of expecting [null pointers](https://en.wikipedia.org/wiki/Null_pointer#C "Description for the null pointer") to be identical to zeros.
* Would you like to add another check for a memory allocation failure here?

**Comment by nikias** [comment]
Yes I will add a check for failing allocation.
Can you make a suggestion on how to change the code? Remember that other code, like collection_add will check for !val[x] so it expects it to be zero. Any idea how to change that in a portable and elegant way?

**Comment by elfring** [comment]
>… !val[x]

This can be also a portable check for a null pointer.


>so it expects it to be zero.

I suggest to reconsider this software interpretation.

**Comment by nikias** [comment]
Can you just suggest how to change the code please.

**Comment by elfring** [comment]
Would you like to take another look at [information from an article like “How to initialise structures to all-elements-zero-or-null”](http://www.ex-parrot.com/~chris/random/initialise.html "Article by Chris Lightfoot")?

**Comment by nikias** [comment]
Fixed with 2501f51e0fe2fad17ce4966542138d3f32dd533c.
